[Step1]인수 테스트 기반 리팩터링 PR 요청 드립니다.

re-mora opened the pull request:
안녕하세요 :) 교육생 최용석입니다 😄

1단계 `인수 테스트 기반 리팩터링` 미션 구현이 완료되어, 리뷰 요청드립니다.

프로젝트에 작성되어 있는 인수테스트 및 기능단위 TC를 기반으로 `LinsService`에 작성된 노선&구간 관련 기능을 연관된 도메인으로 분산하였습니다.

미흡한 부분이 있다면 피드백 부탁드립니다 :)

re-mora committed:
docs: 요구 사항 정리 및 TODO List 작성

re-mora committed:
refactor: 서비스 계층의 구간 추가 기능을 도메인 계층으로 단순 이동

re-mora committed:
refactor: 구간 일급 컬렉션을 활용한 각 도메인 계층으로의 비지니스 로직 분산

구간 일급 컬렉션 정의
- 노선 도메인으로 이동했던 구간 추가 기능을 구간 일급 컬렉션 도메인에 위임
- 신규 구간 시, 기존 구간 변경 로직을 구간 도메인에 위임

노선 기본생성자의 접근제한자를 JPA 사용을 위한 최소 범위로 변경
- AS-IS : public
- TO-BE : protected

접근 제한자 변경으로 인한 서비스 계층 TC 코드 일부 수정
- AS-IS : new Line()
- TO-BE : new Line(name, color)

re-mora committed:
refactor: 원시 타입 포장 객체와 정적 팩토리 메소드 적용

- 원시 타입 포장 객체를 이용한 객체의 유효값 검증 및 관련 로직 위임
- 정적 팩토리 메소드를 이용한 객체 생성 유도 및 생성 시점에 유효성 검증

re-mora committed:
refactor: 리팩토링으로 인해 더 이상 참조가 없는 서비스 계층의 구간 추가 기능 삭제

re-mora committed:
chore(test): 구간 추가 관련 테스트 코드 정리

re-mora committed:
refactor: 서비스 계층의 구간 제거 기능을 관련 도메인으로 분산

re-mora committed:
refactor: 리팩토링으로 인해 더 이상 참조가 없는 서비스 계층의 구간 제거 코드 삭제

re-mora committed:
refactor: 신규 노선 엔티티 및 노선 생성 응답 객체 생성 부를 요청/응답 객체로 위임

re-mora committed:
refactor: 노선 목록 조회 응답 객체 생성 시, 비효율적은 stream 순회 부 수정

re-mora committed:
refactor: 특정 노선 정보 응답 객체 생성 부를 관련 객체로 위임

re-mora committed:
refactor: 노선 정보 수정 시, 노선 엔티티 생성 부를 요청 객체에 위임

re-mora committed:
refactor: 노선 관련 기능을 관련 도메인 객체로 분산함에 따라 더 이상 참조가 없는 코드 제거

re-mora committed:
refactor: 하드코딩 된 예외 처리 문구 상수화 및 코드 정리

re-mora committed:
refactor: 명확하지 않은 변수명 수정

dhmin5693 commented on docs/step1-README.md:
```diff
@@ -0,0 +1,49 @@
+### TODO List
+`LineService`의 기능 목록화
+- 노선 관련 기능
+  - [ ] 노선 저장
+  - [ ] 노선 목록 조회
+  - [ ] 특정 노선 조회
+  - [ ] 노선 정보 수정
```
완성된 기능은 체크해주시면 좋을 것 같아요!

re-mora replied:
문서를 꼼꼼하게 정리하지 못했네요..
중간 중간 작성한 내용을 지속적으로 문서에 반영하는 습관이 아직 부족한 것 같습니다!
말씀해주신 부분은 수정 하도록 하겠습니다!

dhmin5693 commented on src/main/java/nextstep/subway/line/dto/LineResponse.java:
```diff
@@ -30,6 +31,13 @@ public static LineResponse of(Line line, List<StationResponse> stations) {
         return new LineResponse(line.getId(), line.getName(), line.getColor(), stations, line.getCreatedDate(), line.getModifiedDate());
     }
 
+    public static LineResponse of(Line line) {
```
스태틱 팩토리 메소드에서 인자가 1개인 경우 of보다는 관례적으로 from을 사용합니다.
(출처: 이펙티브 자바 item 1)

choi-ys replied:
말씀해주신 내용을 수정하면서 아래와 같이 정리해보았습니다.
의도하신 내용이 맞는지, 제가 적절하게 이해 했는지 궁금합니다!
 
---
스태틱 팩토리 메소드에 위치한 유효성 검증 부를 생성자로 이동
  - 스태틱 팩토리 메소드가 조건에 따라 하위 타입으로의 구현체를 선택하거나, 여러 스태틱 팩토리 메소드가 동일한 생성자를 바라보는 경우 유효성 검증부가 중복 될 수 있다.
  - 여러 스태틱 팩토리 메소드가 가지는 의도에 반하는 경우 각 스태틱 팩토리 메소드는 적절한 유효성 검증 부를 가질 수 있지만
  객체 생성시 필요한 유효성 검증 부는 생성자가 가지도록 하는 것이 조금 더 적절한 위치라고 생각해 볼 수 있다.

re-mora replied:
조금씩 더듬 더듬 공부하고 참고했던 내용들이, 의도는 잊어버리고 무의식적으로 작성하게 된 것 같아요 😭 
반성하면서, 아래와 같이 관련 내용을 한번더 숙지하고 수정하였습니다.

---
 - 스태틱 팩토리 메소드에서 `1개`의 인자를 받아 해당 타입의 인스턴스를 반환하는 경우 : from
 - 스태틱 팩토리 메소드에서 `여러개`의 인자를 받아 해당 타입의 인스턴스를 반환하는 경우 : of

dhmin5693 commented on src/main/java/nextstep/subway/line/domain/Section.java:
```diff
@@ -22,16 +32,63 @@ public class Section {
 
-    private int distance;
+    @Embedded
+    private Distance distance;
```
👍

dhmin5693 commented on src/main/java/nextstep/subway/line/domain/Section.java:
```diff
@@ -22,16 +32,63 @@ public class Section {
+        return new Section(line, upStation, downStation, distance);
+    }
+
+    private static void validateStation(Station upStation, Station downStation) {
+        validateIsNotNullStation(upStation, downStation);
+        validateEqualsStations(upStation, downStation);
+    }
```
이 로직은 팩토리 메소드보다는 생성자에서 수행하는게 위치상 더 적절할 것 같아요.
팩토리 메소드는 변수를 해석한 뒤(Line 등) 생성자로 연결해주는 정도가 제일 깔끔하더라구요.

re-mora replied:
말씀해주신 내용을 수정하면서 아래와 같이 정리해보았습니다.
의도하신 내용이 맞는지, 제가 적절하게 이해 했는지 궁금합니다!

---
### 스태틱 팩토리 메소드에 위치한 유효성 검증 부를 생성자로 이동
- 스태틱 팩토리 메소드가 조건에 따라 하위 타입으로의 구현체를 선택하거나, 여러 스태틱 팩토리 메소드가 동일한 생성자를 바라보는 경우 유효성 검증부가 중복 될 수 있다.
- 여러 스태틱 팩토리 메소드가 가지는 의도에 반하는 경우 각 스태틱 팩토리 메소드는 적절한 유효성 검증 부를 가질 수 있지만
위의 상황에서 객체 생성 시 필요한 유효성 검증 부는 생성자가 가지도록 하는 것이 조금 더 적절한 위치라고 생각해 볼 수 있다.

dhmin5693 replied:
넵 맞습니다.
이게 절대적인건 아니고 저 한명의 개인적인 시각이라는 점, 참고해주세요 :)

dhmin5693 commented on src/main/java/nextstep/subway/line/domain/Section.java:
```diff
@@ -22,16 +32,63 @@ public class Section {
+    }
+
+    public void update(Section section) {
+        if (upStation == section.upStation) {
+            updateUpStation(section.downStation, section.getDistance());
+        }
+        if (downStation == section.downStation) {
```
동등성 검증은 ==가 아니라 재정의한 equals를 사용해야 하지 않을까요?

dhmin5693 commented on src/main/java/nextstep/subway/line/domain/Sections.java:
```diff
@@ -0,0 +1,145 @@
+
+    public List<Station> getAllStation() {
+        if (elements.isEmpty()) {
+            return Arrays.asList();
```
```suggestion
            return Collections.emptyList();
```
이건 어떤가요?

re-mora replied:
비어있는 List 반환 시 가변 객체가 아닌 불변 객체를 반환하도록 수정하였습니다!

---
    - AS-IS : Arrays.asList()
    - TO-BE : Collections.emptyList()

dhmin5693 commented on src/main/java/nextstep/subway/line/domain/Sections.java:
```diff
@@ -0,0 +1,145 @@
+import nextstep.subway.station.domain.Station;
+
+@Embeddable
+public class Sections {
```
이 클래스에 대한 테스트가 없습니다.
이 미션은 ATDD 와 TDD를 같이 체험해보는게 목적입니다.
public으로 외부에 공개되는 로직이 있으면 반드시 테스트를 작성해주시고 충분한 만큼 검증까지 해주세요 :)

dhmin5693 commented on src/main/java/nextstep/subway/line/application/LineService.java:
```diff
@@ -8,19 +10,13 @@
 @Service
-@Transactional
+@Transactional(readOnly = true)
 public class LineService {
```
서비스 레이어에 대한 단위 테스트도 작성해주세요.
Mockito를 활용해주시면 됩니다.

dhmin5693 commented on src/main/java/nextstep/subway/line/domain/wrap/Distance.java:
```diff
@@ -0,0 +1,53 @@
+
+@Embeddable
+@Access(AccessType.FIELD)
+public class Distance {
```
이 클래스도 검증이 필요합니다.

dhmin5693 commented:
전반적으로 테스트가 작성되어 있지 않아요.
이번 미션은 TDD를 함께 경험하는 것이므로 테스트를 꼼꼼하게 작성해주시면 감사하겠습니다!

dhmin5693 reviewed (changes requested):
안녕하세요 용석님!
ATDD-TDD 미션을 함께할 민동휘라고 합니다.
이번 미션에서 많은 것을 얻어가실 수 있도록 성심성의껏 지원해드리도록 하겠습니다. 🙇‍♂️
혹시라도 제가 남기는 피드백이 이해가 안된다거나 납득이 안되는 부분이 있으면 언제든지 편하게 DM 보내주세요!

미션에 대한 코멘트를 몇 개 남겨두었으니 확인 후 재요청 부탁드립니다. 🙏

re-mora committed:
docs: 코드 리뷰 피드백 내용 정리

- 스태틱 팩토리 메소드 활용에 관련된 리뷰 및 피드백 적용을 위한 내용 정리
- 작성된 프러덕션 코드에 대해 작성 해야하는 TC 목록

re-mora committed:
docs: 명확하게 표현되지 않은 TODO List 수정

re-mora committed:
refactor: 스태틱 팩토리 메소드에 위치한 유효성 검증 부를 생성자로 이동

re-mora committed:
refactor: 비어있는 List 객체 반환 부 수정

AS-IS : Arrays.asList() -> new ArrayList<>()
  - 가변 객체를 반환
TO-BE : Collections.emptyList() -> EmptyList<E>
  - 불변 객체를 반환

re-mora committed:
refactor: 객체 동등성 비교 시, `==`로 작성된 코드를 `equals`로 변경

re-mora committed:
test: LineService 기능에 대한 TC 작성

re-mora committed:
test: Distance에 대한 TC 작성

re-mora committed:
test: 일급 컬렉션인 Sections에 대한 TC 작성

re-mora commented:
안녕하세요 동휘님 :)
리뷰 해주신 내용을 수정해 보았습니다.
말씀하셨던 부족한 테스트를 작성 위주로 커밋하였습니다.
미흡한 부분이 있다면 한번더 피드백 부탁드리겠습니다!

re-mora committed:
chore(re-name): `1개`의 인자를 받아 해당 타입의 인스턴스를 반환하는 스태틱 팩토리 메소드 명 수정

AS-IS : of
TO-BE : from

dhmin5693 reviewed (approved):
피드백 잘 반영해주셨습니다.
다음 미션 진행해주세요.
수고하셨습니다.
